### DISCUSSION TITLE
Remote control and peak shaving

### DISCUSSION THREAD

**Discussion by johanymerson** [wills106/homeassistant-solax-modbus#283] [Ideas]
Hi,

First of all, thanks for an excellent  work on this HA plugin!
I have been testing the new remote control functions, and from what I can see all three modes work as intended on my X3 Gen4. With today's volatile electricity prices this is very useful.
With a small modification I think this could also be used for Peak Shaving, by using the Grid Control mode and setting active power to what you want to limit your grid import to. To be useful for Peak Shaving the Active Power positive limit would need to be removed when in Grid Control mode.

/Johan

**Comment by infradom** [comment]
Thanks for the comment, we indeed have the plan to implement another computed mode similar to "Enabled Grid Control", but intended to just determine the max grid import. It should not be that difficult to implement, we only wanted to make sure the currently implemented modes work well, in the presence of PV as well as without PV.
Do you also need peak shaving for export? (probably yes, on days with negative export prices)

**Reply by johanymerson** [reply]
We haven't had to deal with negative export prices here yet, but I wouldn't rule it out.
I tested all three modes yesterday, with and without PV. From what I could see all three modes behaved as expected.
I did a quick test today and just remove the max value for Active Power, then set it to 10kW and selecting Grid Control mode. That did indeed cap the import to 10kW, and the battery was charged between the peeks to 100%. That's of course not a good solution, more a POC.
![image](https://user-images.githubusercontent.com/2807771/210005806-d61a648b-dd66-4d70-ac5f-8a82611f49a9.png)

**Comment by infradom** [comment]
Thanks, but I am not sure what you mean by remove the max value for Active Power.
Did you modify the source code that keeps the active power in the range indicated by the inverter:

> ` line 122 in plugin_solax.py           'remotecontrol_active_power':   max(min(ap_up, target),   ap_lo),`

Or do you meane something else?
My understanding of peak shaving is the usecase where the inverter uses its power from the grid (and leaves battery idle), but in case grid import exceeds for example 3kW, it takes the excess load from the battery. Of course there are variations with PV and a non-completely idle battery.

In my Flemish Belgian region, from Januari 1st 2023 on, we get taxed for peaks (15 minute intervals), so it is important for me too to come up with a solution that avoids grid import peaks.
In the cases below, for simplicity, I assume there is no PV:
- As long as there is sufficient PV or high battery levels, there will be no import at all in normal self use mode.
- Sometimes "Enabled Grid Control" mode is used to charge the battery at a well defined grid import power level
- There is a "Enabled Battery Control" mode that can be used  when target is set to zero to let the grid deliver the house load (without import limits)

I would like to implement a variation of that Battery Control, where the target value is more a maximum import level, and where the battery  target charge/discharge is zero (or configurable) as long as we do not exceed the grid import max level.

Your picture around 20:40 seems to show  a situation that matches such a situation, but I may not understand it completely

**Reply by johanymerson** [reply]
My use-case here is similar to yours, our grid companies also have started charging based on the highest peak hour instead of by the main fuse size, so I want to limit the peak grid import by both shaving peaks and controlled charge (we don't have enough PV power this time of year to export anything).
I cleaned up my changes and pushed them here so you can have a look:
https://github.com/johanymerson/homeassistant-solax-modbus

In Grid Control mode the target power (battery in/out) is calculated like this:
target = target - (datadict['inverter_load'] - datadict['measured_power'])
That means that the target configured has an offset of measured_power from active power. In the other modes the Active Power should be limited to inverter capacity (model dependent) or installed battery capacity (number of battery packs installed, each pack can supply 2.5kW). But in Grid Control mode it shouldn't be be limited as the house load can be much higher than the inverter capacity.
Since it would be difficult to implement different limits of remotecontrol_active_power depending on mode I created a separate variable for Grid Control mode instead, remotecontrol_grid_power.
 
To use it for peak shaving (limit import), set the maximum import power you want in Remotecontrol Grid Power, say 10000W, and enable Grid Control mode. That way it will use the battery when the import power exceed 10kW, and charge the battery when the import power is lower than 10kW.

**Reply by infradom** [reply]
You might be right that the house_load might be significantly higher than the inverter power, so the idea of introducing a new variable makes sense. I will give it some thought (the name should not be confusing). Be aware that the MAX_EXPORT levels defined by @wills106 are already significantly higher than the inverter's theoretical power.

**Comment by infradom** [comment]
One more remark: the remotecontrol active power limits are now restricted to 6000W. This limit should become dependent on the inverter power, so I suggest we modify that in the coming days

**Reply by johanymerson** [reply]
Yes, I was about to comment on that. But that is not trivial as the actual limit will depend on the number of battery packs installed, 2.5kW per pack for the Triple Power 3.0. I'm not sure it is possible to get the actual number of installed packs over modbus so you might need to guess that from the battery_voltage_charge.

**Reply by infradom** [reply]
Thanks !
I have enhanced the wiki page with a couple of example automations.
On my local github desktop, I already made the maximum import/export levels dependent on the type of inverter.
I will soon issue a PR with these changes, so that @wills106  can review them, but it will probably not be for today as I want to release it together with the new peak_shaving mode.
The batteries may be different, but it is up to the user to configure the automations accordingly.

**Reply by johanymerson** [reply]
Yeah, I think that is the best approach. I had a quick look and it would be quite difficult to estimate how much the battery actually will be able to supply (or how much charge it will accept). On top of that it will be highly temperature dependent.

**Comment by infradom** [comment]
One more remark:
I dont consider the current grid_control mode the peak_shaving mode I try to implement
The grid_control mode will always import the specified target.
I would like to add a mode where the import is normally equal to the house load, but not exceeding a specified grid_import_max
So the grid import will vary continuously with the house load, but will be 'shaved' if it exceeds the max

**Reply by johanymerson** [reply]
Ok, but in peak shaving mode the battery need to be topped up between the peaks. If you put it in grid control mode with my modifications it will top up the battery to 100% between the peaks, but that does not mean it will keep the import at the configured limit all the time. When the battery is full there is nowhere for the power to go, so then import will be equal to house load.
Have a look at the graph above. At 20:38 the load exceeded the configured 10kW and the inverter kicked in and supplied the excess power from battery. When the house load dropped below 10kW (at 20:43) the inverter topped up the battery again and then idled (house load = import).

**Comment by infradom** [comment]
I am less convinced that the battery (always) needs to be topped up from grid. If I expect a lot of sunshine in the coming hours, I do not want to top up the battery with more expensive grid power. I just want to keep the current battery level until the PV power is there, grid provides the house load in this situation, but needs to be shaved
It is confusing, since the summer and winter patterns are completely different ...

**Reply by johanymerson** [reply]
Haha, yes, it is 3 months since I had any excess PV power to charge the battery with. PV yield today was 1.8 kWh, and this was a good day... 
What about a Peak Shave mode with configurable minimum SOC? When the SOC is above this level the (actual) target value is limited to 0 to prevent charge from grid. Below this level the the target value is set to charge the battery from grid slowly (probably also configurable).
If you know you have a lot of PV power you can set the min SOC level to something low, just so you survive a peak in the early morning. If you, like me, don't have any excess PV power you set it to 90-100%.

**Reply by infradom** [reply]
Sorry, a week ago,  there were some sunny days on which I still could charge 13kWh/day into my batteries, so not every installation and location is the same.
In Belgium, with a reasonable amount of panels,  really dark periods in which PV can be ignored are very rare and occur only in december and january

**Reply by johanymerson** [reply]
To show what I mean I did a quick POC and pushed here:
https://github.com/johanymerson/homeassistant-solax-modbus/commit/57c17d818623393e369f87a2220d1d9f16c97ef1
I have done basic testing and it does appear to work, but I don't have any PV power at the moment.

**Reply by infradom** [reply]
Thanks for tue suggestion, but I am still searching another solution. I believe the autorepeat loop should only be used to react to fast changing load variations. Slow changing things like SOC belong to the customer's automation strategies, not to the integration. At least this is my opinion.
So I just feel we need is new mode that just shaves import peaks, while the grid import follows the house load as long as the import level does not exceed the defined max.
Then I would use automations to switch between modes at a slow rate (e.g. twice a day), these automation use soc,  solar forecast and load forecast  to determine the strategy.

**Reply by johanymerson** [reply]
I have modified the code now to be a pure peak shaving mode with no charging. Is this what you mean?
https://github.com/johanymerson/homeassistant-solax-modbus/commit/74c7fb87e312f5236106a0f2c8c7a4c2e457e95f
My fear though is that it might get difficult to do the battery charging separate in an automation. We need peak shaving to be active all the time, even during charging, so there need to be an automation that flip the mode back to peak shave as soon as grid import exceed the import limit. Since the Battery or Power control modes doesn't do shaving they can exceed the limit, thus triggering a flip back to Peak Shave. By having the charging ability built into the Peak Shave mode that can be avoided.

**Comment by johanymerson** [comment]
I noticed in the wiki you added an example automation to force charging when prices are low.
I might have misunderstood how the Grid Control mode is supposed to work (and/or the Grid Control mode is not working as intended). When I test that example, setting Active Power to 2500 W and selecting Enabled Grid Control, my battery is **discharging**, because my house load - PV power is higher than 2500 W. To use Grid Control mode to charge my battery I would need to set Active Power to something like 11500 W, as my house load is about 8000 W at the moment. But that is higher than the maximum limit of the Active Power local register.
To charge the battery slowly independent on house load I have used Enabled Battery Control and Active Power = 2500 W successfully.

**Comment by infradom** [comment]
On my github development fork, https://github.com/infradom/homeassistant-solax-modbus I now have implemented a version with 2 new controls:

- remotecontrol_import_limit
- remotecontrol_export_limit

I dont need an extre mode to implement peak shaving with this extension as it applies to all modes (but only makes sense with "Enabled Power Control" and "Enabled_Battery_Control"
This new version also makes the limits dependent on your inverter type

My automations tend to use 

- "Enabled Power Control" when prices are at a reasonably low level and dont care about charging battery or not
- "Enabled Grid Control" when prices are extremely low and I definitely want to charge the battery at desired speed

Sometimes I leave remote power control disabled

**Reply by johanymerson** [reply]
I don't understand how that these new controls are supposed to be used. They are limited to the the inverter maximum power, for example 10kW. But my grid import (house load - PV) can be much higher than the inverter power rating. So to use it for peak shaving I need to be able to set a value higher than the inverter maximum power.
How would I set peak shaving up using these controls if my inverter maximum is 10kW and I want peak shaving at 12kW import power?

**Comment by infradom** [comment]
In the current implementation, although my inverter is only 8 kW, we allow grid import/export up to 12 kW
For your 10kW Gen4, I believe the current limit will be 15 or 16 kW. See MAX_EXPORT list in plugin_solax.py
We can easily increase these limits, @wills106 has chosen reasonable limits for each inverter type. 
The only reason for limiting them a bit is to make the sliders more useable.

**Comment by infradom** [comment]
The first implementation was intended for initial tests, and had this fixed 6000 watt limit. The development version in my development fork already uses limits that are inverter dependent and typically 150% of the inverter power. We can increase that if there is a need or make it configurable, but maybe the 150% already meets your needs

**Reply by johanymerson** [reply]
I'm not sure having limits on the import/export limits or in grid control mode makes sense. Inverter size is chosen based on installed PV power, not on house load. So it is quite common to have inverters much smaller than the potential peak import. 3*25 A is probably the most common electrical service here, meaning peak powers in the range of 15-20 kW, and at the same time have a 5-10kW inverter.

Also, since peak shaving is not a separate mode, and the import limit is limited, in my case to 15 kW, there is no way for me now to disable peak shaving. It will now cap my import to max 15kW and thus limit my possibilities to charge from grid during the cheapest hours. (Some grid companies only count peaks during high-price hours, so exceeding the limit night time is fine)

**Reply by infradom** [reply]
I think we need to solve your limit problem, no discussion about that
We seem to have completely different opinion on how power management needs to be done. For that reason, this part should be in the automations or an external EMS system like emhass (I am working on that)
In my opinion, this integration only needs to provide some simple working modes that makes it possible to implement peak shaving (whatever people may understand under that term; alhough i have a 3x40 Amp household, and 8kW inverter, I intend to limit my imports to 2500 Watt to avoid net usage charges)

**Comment by infradom** [comment]
In my development fork, I have made the limits on remotecontrol_import_limit and export_limit 30kW fixed
(For export this was not really needed as it depends on the inverter power, but I did it to make both controls look similar)
 https://github.com/infradom/homeassistant-solax-modbus 
I think the export limiting does not work correctly yet (but I need more PV power to test in depth).
I will keep this in my development branch a while, I want to hear feedback and @wills106 opinion before issuing a PR
The documentation of the 2 new controls is also still missing in the wiki, but I believe it explains itself

**Reply by johanymerson** [reply]
I started testing this today but the solax component crashed with:
```
Traceback (most recent call last):
  File "/var/lib/homeassistant/.homeassistant/custom_components/solax_modbus/__init__.py", line 404, in read_modbus_data
    res = self.read_modbus_registers_all()
  File "/var/lib/homeassistant/.homeassistant/custom_components/solax_modbus/__init__.py", line 485, in read_modbus_registers_all
    if self.localsUpdated: self.saveLocalData()
  File "/var/lib/homeassistant/.homeassistant/custom_components/solax_modbus/__init__.py", line 224, in saveLocalData
    with open(f'{self.name}_data.json', 'w') as fp: json.dump(tosave, fp)
PermissionError: [Errno 13] Permission denied: 'SolaX_data.json'
```

I assume you want the new .json file to end up in the config directory so I modified __init__.py like this:
```diff
diff --git a/custom_components/solax_modbus/__init__.py b/custom_components/solax_modbus/__init__.py
index 86cfebf..d7675c3 100644
--- a/custom_components/solax_modbus/__init__.py
+++ b/custom_components/solax_modbus/__init__.py
@@ -221,12 +221,12 @@ class SolaXModbusHub:
     def saveLocalData(self):
         tosave = { '_version': self.DATAFORMAT_VERSION }
         for desc in self.writeLocals:  tosave[desc] = self.data.get(desc) 
-        with open(f'{self.name}_data.json', 'w') as fp: json.dump(tosave, fp)
+        with open(self._hass.config.path(f'{self.name}_data.json'), 'w') as fp: json.dump(tosave, fp)
         self.localsUpdated = False
         _LOGGER.info(f"saved modified persistent date: {tosave}")
 
     def loadLocalData(self):
-        try: fp = open(f'{self.name}_data.json') 
+        try: fp = open(self._hass.config.path(f'{self.name}_data.json')) 
         except: pass
         else: 
             loaded = json.load(fp)
```

I'll experiment with automating the peak shave mode (jumping between peak shave and slow charge between peeks) later today, but so far it looks good :thumbsup:

**Reply by infradom** [reply]
Thanks, I will adapt the code !
As you probably have understood, either you can work with

- Grid Control mode, and then the import limit does not make much sense as the target is aready exactly defined
- Power Control mode, with a certain active_power°target,  in which the import_limit  can put a limit on the grid import
Export limit did not work as expected yet for me

Thanks for your suggestions !

**Reply by infradom** [reply]
Just for your information: today, electricity prices were extremely low (almost for free), so my batteries were loaded close to 100% and I ran the entire day in "Enabled Power Control" mode with active_power target=0 with an import limit=2500W
So when the PV was low, most of the house load came from the grid, and my batteries remain full for tomorrow or later, when we expect higher prices

**Reply by johanymerson** [reply]
Prices were really low here too today. If it wasn't for the tax they would have been negative. So an excellent day to play with the inverter settings.
I did test the import limit with both Power Control and Battery Control. Both of them seem to be doing what they should, both with and without PV power (the little we have). Really looking forward to some sunlight again, the sun doesn't even reach 10° above the horizon at the moment, and is only up for 4 hours...
I have set up an automation that enables Battery Control with import limit if import power goes above that level, and starts slow charging from grid if the import power go significantly below that level for an hour. My only concern is that these automations could result in an oscillation, but hopefully the long delay (1 hour) and hysteresis will prevent that.

**Reply by infradom** [reply]
Although the current solution works  well for me, I still have the feeling we could even simplify things. The power_control mode combined wiith import_limit is very similar to grid control mode. 
So I believe Grid_Control(target=4000W) is the same as Power_Control(target=4000W, import_limit=4000W)
So if this observation is right, Grid_Control mode is not really needed as it is less general than Power_Control mode. It is simpler to understand however.
I might be wrong.. I must admit this confuses me a lot

**Reply by johanymerson** [reply]
Yeah, it the moment it is a bit difficult to understand all aspects of these modes.
Let's start with the basic modes, without the import/export limits. This is how I understand it:
![image](https://user-images.githubusercontent.com/2807771/210256231-45315032-8436-41b3-84c1-0a7b6cf102ce.png)
How excess solar power is used I'm not sure, as I don't have any excess solar panel. Say you have configured:
Power Control (target=-10000, import_limit=30000,export_limit=0) to simulate "Self-use during discharge hours", and the solar power exceed the house load. Will export still be limited to 0 then (i.e. solar throttled), or will the export limit be exceeded?
And how does this compare to:
Power Control (target=0, import_limit=0, export_limit=30000), which in my tests also behave as "Self-use during discharge hours"?
If you have excess solar power tomorrow could you test both of these and report back?

**Reply by infradom** [reply]
I have the impressin that with this mechanism, the solar is not throttled, so export limit did not work last time I tested this.
There may be an error in my formula; I will check further tomorrow, if it is sunny.
Until now, the only way I could cut export was by changing the export_control_user_limit. But this parameter is written to eeprom, so we need to be careful not to change it too often.
I am convinced there must be another way, problem is that I have no official documentation at all about the solax modbus power control mechanims

**Reply by infradom** [reply]
your drawing is correct,, but you could also add a 'Power-Control' line that covers house load and grid  (as explained in the updated wiki):
```
Conclusion: we can look at Activepower in 2 ways (both ways are correct):

Activepower = grid(_import) - house_load
Activepower = battery(_charge) - PV
```

**Reply by johanymerson** [reply]
I have been testing a number different use cases and haven't found one that required the Grid Control mode. I saw in your fork I see you also removed the export_limit. I agree that it is somewhat misleading and not really useful. Export control is more of an installer setting to comply with local regulations anyway. What you may want to limit is "discharge to grid", similar to the Self Use mode during discharge hours. That can be done by using Power Control (active_power=0, import_limit=0).

On thing though, when I test your current development fork I get the following error when hitting trigger button:
```
Traceback (most recent call last):
  File "/var/lib/homeassistant/lib/python3.10/site-packages/homeassistant/components/websocket_api/commands.py", line 200, in handle_call_service
    await hass.services.async_call(
  File "/var/lib/homeassistant/lib/python3.10/site-packages/homeassistant/core.py", line 1745, in async_call
    task.result()
  File "/var/lib/homeassistant/lib/python3.10/site-packages/homeassistant/core.py", line 1782, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/var/lib/homeassistant/lib/python3.10/site-packages/homeassistant/helpers/entity_component.py", line 213, in handle_service
    await service.entity_service_call(
  File "/var/lib/homeassistant/lib/python3.10/site-packages/homeassistant/helpers/service.py", line 678, in entity_service_call
    future.result()  # pop exception if have
  File "/var/lib/homeassistant/lib/python3.10/site-packages/homeassistant/helpers/entity.py", line 943, in async_request_call
    await coro
  File "/var/lib/homeassistant/lib/python3.10/site-packages/homeassistant/helpers/service.py", line 715, in _handle_entity_call
    await result
  File "/var/lib/homeassistant/lib/python3.10/site-packages/homeassistant/components/button/__init__.py", line 116, in _async_press_action
    await self.async_press()
  File "/var/lib/homeassistant/.homeassistant/custom_components/solax_modbus/button.py", line 85, in async_press
    res = self.button_info.value_function(0, self.button_info, self._hub.data )
  File "/var/lib/homeassistant/.homeassistant/custom_components/solax_modbus/plugin_solax.py", line 139, in value_function_remotecontrol_recompute
    _LOGGER.debug(f"peak shaving: old_ap_target:{old_ap_target} new ap_target:{ap_target} max: {import_limit-houseload} min:{-export_limit-houseload}")
NameError: name 'export_limit' is not defined
```

**Reply by wills106** [reply]
Just so you know 'export_limit' was removed just before the changes were pushed from @infradom back to me.
So you might have a partial setup?

You are probably best updating to the latest stable / beta on my Main branch.

**Reply by johanymerson** [reply]
Problem seem to be the same in your main branch. Traced it down to a debug printout still using the export_limit:

```diff
diff --git a/custom_components/solax_modbus/plugin_solax.py b/custom_components/solax_modbus/plugin_solax.py
index 076989a..06f050f 100644
--- a/custom_components/solax_modbus/plugin_solax.py
+++ b/custom_components/solax_modbus/plugin_solax.py
@@ -136,7 +136,7 @@ def value_function_remotecontrol_recompute(initval, descr, datadict):
     ap_target = min(ap_target,  import_limit - houseload)
     #_LOGGER.warning(f"peak shaving: old_ap_target:{old_ap_target} new ap_target:{ap_target} max: {import_limit-houseload} min:{-export_limit-houseload}")
     if  old_ap_target != ap_target: 
-        _LOGGER.debug(f"peak shaving: old_ap_target:{old_ap_target} new ap_target:{ap_target} max: {import_limit-houseload} min:{-export_limit-houseload}")
+        _LOGGER.debug(f"peak shaving: old_ap_target:{old_ap_target} new ap_target:{ap_target} max: {import_limit-houseload}")
     res = { 'remotecontrol_power_control':  power_control,
             'remotecontrol_set_type':       set_type,
             'remotecontrol_active_power':   max(min(ap_up, ap_target),   ap_lo),

```

**Reply by wills106** [reply]
When you changed code did you remove the Integration from the Integration page and restarted HA before adding the Integration back in?

Unfortunately there are situations where if you don't remove the Integration before upgrading and restarting and adding the Integration again you get strange behaviour from old and new code being present in HA.

I am just wondering if you had removed and readded the Integration, would the Integration even know about 'export_limit'

I am sorry if that doesn't make any sense.

**Reply by johanymerson** [reply]
No, didn't do that in this case. I have done in the past to clean up strange remains in the HA database. But in this case the problem is still in the code. Looking at the current state of your main:
![image](https://user-images.githubusercontent.com/2807771/210876224-d7f4a7c0-e19e-4cb0-8783-cab523386e21.png)

If you hit this code or not may depend on how you have your debug levels set.

**Reply by wills106** [reply]
I can't debug this exact code as I don't own a Gen4... But you can get into odd situations where phantom bits of code stuck in HA can trigger stuff that wouldn't be possible from a fresh installation.

Hopefully @infradom can verify and clean up when he's next free over the next couple of days.

For now if you do a clean install does the issue go away?

**Reply by johanymerson** [reply]
No, but applying the patch I pasted above does. The problem is that export_limit is still being referred to in the debug printout. If you have logging set to log debug printouts it will execute and fail. If you don't have debug logging enabled the code won't execute.

**Comment by infradom** [comment]
Thanks and Sorry, I corrected this and issued a PR, as I am currently not in debug mode, this error did not appear on my system(afaik)

**Comment by infradom** [comment]
For your information: after some tests, I found out that remotecontrol_export_limit was not capable of throttling the PV power, so this parameter became useless. In cases op periods with negative prices, I have an automation that sets export_control_user_limit to zero.
(these periods do not occur that often these days, so for the time being, we dont need  to worry about the EEPROM writes associated with this command)

**Reply by johanymerson** [reply]
Yeah, I suspected it wouldn't be able to throttle the PV power. Negative prices haven't been an issue here yet, not including tax return anyway. They have come close but only night time, and there is no PV power to worry about then.

**Reply by LadCZ** [reply]
How is it with EPROM writes in Solax X3 G4? Wiki documentation of activepower mentions, there are no writes. Do you mean only the active power entities or nothing is written into EPROM on G4?

**Reply by infradom** [reply]
> How is it with EPROM writes in Solax X3 G4? Wiki documentation of activepower mentions, there are no writes. Do you mean only the active power entities or nothing is written into EPROM on G4?

We perform writes to the inverter, but these write operations are not written to EEPROM as they are only valid for a short period (x seconds). The Solax documentation also states that these write operations are not written to EEPROM, unlike other parameters.

**Reply by infradom** [reply]
Additional comment: The remote_control parameters are saved in  a HA config file, not in the inverter.

**Reply by LadCZ** [reply]
So I don't need to worry about demanding EPROM by excessive number of commands sent to solax? That woud be great. Can you please share to domumentation?

**Reply by infradom** [reply]
Not as long as you use the remotecontrol_xxx commands documented on the wiki page: https://github.com/wills106/homeassistant-solax-modbus/wiki/Modbus-Power-Control-(remotecontrol-power)
For a Gen3 inverter, this may be slightly different/
Unfortunately, I cannot share the documentation

**Reply by LadCZ** [reply]
Ok thanks.

**Reply by iMatte81** [reply]
Hi, are there some news about Gen3 inverter and remotecontrol?
Thank you

**Reply by wills106** [reply]
You can Force Charge / Discharge the Inverter following https://github.com/wills106/homeassistant-solax-modbus/wiki/SolaX-Gen2-&-Gen3-Modes-of-operation#grid-export-gen-3-only
You can actually use a positive value to force Charge. The Wiki needs updating.

But the Gen4 style using the write_registers instead hasn't been setup yet.

**Reply by iMatte81** [reply]
Thank you.
